### DISCUSSION TITLE
Unexpected memory allocation

### DISCUSSION THREAD

**Discussion by PsycHead** [slint-ui/slint#6718] [Q&A] [answered]
Memory usage of Rust programs compiled with Slint seems unexpected:
1. Memory does not seem to be freed after dropping all Slint related objects
2. using QT as a back end, initial memory usage is high even before creating any windows

### Example
A program running in the background for a while, then creating a window and eventually closing again.
I made a minimal repo based on the Slint template to reproduce the issue ( https://github.com/PsycHead/slint-memory-test ).

### Expected Behavior
1. The program starts with a fairly low memory footprint (the executable)
2. As the Window is drawn, more memory is allocated
3. After dropping all Slint objects, the allocated memory is released

### Current Behavior
RAM usage on my system (Arch Linux, Wayland) for all three program phases. (Phase - QT - Winit)
1. Startup, no UI - 70MB - 2MB
2. UI visible - 102MB - 167MB
3. UI droppend - 107MB - 167MB

Note that memory is not freed after dropping the UI and that QT immediately increases the memory footprint by a lot.

### Why is this an Issue?

My use case is a service that will run in the background and wait for a signal to occasionally show an interactive window. Consuming more than 100MB for a dormant program is not acceptable in that case.

Discussions #5854 and #3376 may be related. Adding `SLINT_DESTROY_WINDOW_ON_HIDE=1` does nothing on my machine.

**Comment by ogoffart** [comment]
- How do you measure the memory?
- When you use winit, what renderer do you use (skia or femtovg?)

There are several cause of memory usage. Like caches at different level. For example, for the fonts, the images. Or within the GPU driver for some textures. Some of which we have control over, some of which we don't.
I predict that the software renderer is the one using the less memory as it doesn't use the GPU and has almost no caching.

**Comment by PsycHead** [comment]
* I am measuring total allocated physical memory of the process to get a ballpark figure. (The repo uses the [memory_stats](https://docs.rs/memory-stats/latest/memory_stats/) crate, which in turn parses [/proc/self/smaps](https://www.man7.org/linux/man-pages/man5/proc_pid_smaps.5.html) )
* I am using femtovg with both backends. Using Skia makes no noticeable difference. 

It makes sense that different backends on different platforms have different behavior, which is only partially under your control. I also think the memory required when rendering UI is absolutely fair!

The things that surprised me are that just having Slint imported will immediately increase the memory footprint (with QT), even if nothing is drawn or explicitly initialized and that memory is not freed after dropping all objects.

Is there any way to control this behavior?

**Reply by ogoffart** [reply, marked as answer]
Thanks for the extra details!

The increased memory with the Qt backend is likely due to how Qt initializes resources, including a font cache, even if no UI is shown. This is built into Qt to make later UI actions faster, but we don't have control over it.

As for memory not freeing after dropping all Slint objects, that’s due to caching. There’s currently no easy way to fully "unload" them once initialized.

If keeping memory low when idle is crucial, I’d recommend trying the winit backend with the `software` renderer, as it uses less memory overall.

**Reply by PsycHead** [reply]
I see, thank you for the insights!

Since I need a native look for my application, I'll try to separate it into two executables. 
One will be the minimal background process, which will then execute the second application, drawing the UI. If startup time is fast enough, this should be feasible.
